cmd/nerdctl/ipfs_registry_up.go

```diff
@@ -0,0 +1,108 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/containerd/nerdctl/pkg/idutil/containerwalker"
+	httpapi "github.com/ipfs/go-ipfs-http-client"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+)
+
+const ipfsRegistryContainerName = "ipfs-registry"
+
+func newIPFSRegistryUpCommand() *cobra.Command {
+	var ipfsRegistryUpCommand = &cobra.Command{
+		Use:           "up",
+		Short:         "start registry as a background container \"ipfs-registry\", backed by the current user's IPFS API",
+		RunE:          ipfsRegistryUpAction,
+		SilenceUsage:  true,
+		SilenceErrors: true,
+	}
+
+	ipfsRegistryUpCommand.PersistentFlags().String("registry-address", defaultIPFSRegistryAddr, "address to listen")
+
+	return ipfsRegistryUpCommand
+}
+
+func ipfsRegistryUpAction(cmd *cobra.Command, args []string) error {
+	client, ctx, cancel, err := newClient(cmd)
+	if err != nil {
+		return err
+	}
+	defer cancel()
+	walker := &containerwalker.ContainerWalker{
+		Client: client,
+		OnFound: func(ctx context.Context, found containerwalker.Found) error {
+			return nil
+		},
+	}
+	nc, err := walker.Walk(ctx, ipfsRegistryContainerName)
+	if err != nil {
+		return err
+	}
+	if nc > 0 {
+		logrus.Infof("IPFS registry %q is already running. Using the existing one. Restart manually if needed.", ipfsRegistryContainerName)
+		return nil
+	}
+	return runRegistryAsContainer(cmd)
+}
+
+// runRegistryAsContainer runs "nerdctl ipfs registry serve" as a container with --net=host.
+// This function bind mounts nerdctl binary to a directory and runs that directory as the rootfs.
```

Comment:
This doesn’t work when the nerdctl binary is dynamically linked.
In such case this command should print a human-readable error that recommends to use the serve command directly